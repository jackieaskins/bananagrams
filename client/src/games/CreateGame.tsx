import { Box, Button, Heading, Stack } from "@chakra-ui/react";
import { useCallback, useState } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import ErrorAlert from "../alerts/ErrorAlert";
import InputField from "../forms/InputField";
import CenteredLayout from "../layouts/CenteredLayout";
import { useSocket } from "../socket/SocketContext";
import { GameInfo, GameLocationState } from "./types";

export default function CreateGame(): JSX.Element {
  const [gameName, setGameName] = useState("");
  const [username, setUsername] = useState("");
  const [error, setError] = useState("");
  const [isCreatingGame, setIsCreatingGame] = useState(false);
  const navigate = useNavigate();
  const { search } = useLocation();
  const isShortenedGame = new URLSearchParams(search).has("isShortenedGame");

  const { socket } = useSocket();

  const onSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement>): void => {
      event.preventDefault();

      setIsCreatingGame(true);

      socket.emit(
        "createGame",
        { gameName, username, isShortenedGame },
        (error: Error, gameInfo: GameInfo) => {
          if (error) {
            setError(error.message);
            setIsCreatingGame(false);
          } else {
            const locationState: GameLocationState = {
              isInGame: true,
              gameInfo,
            };

            navigate(`./game/${gameInfo.gameId}`, {
              state: locationState,
            });
          }
        },
      );
    },
    [gameName, isShortenedGame, navigate, socket, username],
  );

  return (
    <CenteredLayout>
      <Box as="form" autoComplete="off" onSubmit={onSubmit}>
        <Stack spacing={6}>
          <Heading as="h1" textAlign="center">
            Start a new game
          </Heading>

          <Stack>
            <ErrorAlert visible={!!error}>{error}</ErrorAlert>

            <InputField
              label="Game name"
              name="gameName"
              isRequired
              value={gameName}
              setValue={setGameName}
            />

            <InputField
              label="Username"
              name="username"
              isRequired
              value={username}
              setValue={setUsername}
            />
          </Stack>

          <Button
            colorScheme="blue"
            isLoading={isCreatingGame}
            loadingText="Creating game"
            type="submit"
          >
            Create game
          </Button>
        </Stack>
      </Box>
    </CenteredLayout>
  );
}
