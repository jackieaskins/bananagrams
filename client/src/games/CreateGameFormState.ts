import { useState, MouseEvent } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

import { useSocket } from '../socket/SocketContext';
import { SetState } from '../state/types';
import { GameInfo, GameLocationState } from './types';

export type CreateGameFormState = {
  error: string;
  gameName: string;
  isCreatingGame: boolean;
  onSubmit: (event: MouseEvent<HTMLElement>) => void;
  setGameName: SetState<string>;
  setUsername: SetState<string>;
  username: string;
};

export const useCreateGameForm = (): CreateGameFormState => {
  const [gameName, setGameName] = useState('');
  const [username, setUsername] = useState('');
  const [error, setError] = useState('');
  const [isCreatingGame, setIsCreatingGame] = useState(false);
  const { push } = useHistory();
  const { search } = useLocation();
  const isShortenedGame = new URLSearchParams(search).has('isShortenedGame');

  const { socket } = useSocket();

  const onSubmit = (event: MouseEvent<HTMLElement>): void => {
    event.preventDefault();

    setIsCreatingGame(true);

    socket.emit(
      'createGame',
      { gameName, username, isShortenedGame },
      (error: Error, gameInfo: GameInfo) => {
        if (error) {
          setError(error.message);
          setIsCreatingGame(false);
        } else {
          const locationState: GameLocationState = {
            isInGame: true,
            gameInfo,
          };

          push(`/game/${gameInfo.gameId}`, locationState);
        }
      }
    );
  };

  return {
    error,
    gameName,
    isCreatingGame,
    onSubmit,
    setGameName,
    setUsername,
    username,
  };
};
