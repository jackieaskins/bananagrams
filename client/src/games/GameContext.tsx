import React, { createContext, useContext, useEffect, useState } from 'react';
import { useHistory, useLocation, useParams } from 'react-router-dom';

import { useSocket } from '../SocketContext';
import { SetState } from '../state/types';

type GameParams = {
  gameId: string;
};

export type GameLocationState = {
  isInGame?: boolean;
  gameInfo?: GameInfo;
};

export type Player = {
  userId: string;
  username: string;
  isOwner: boolean;
  isReady: boolean;
};

export type GameInfo = {
  gameId: string;
  gameName: string;
  isInProgress: boolean;
  players: Player[];
};

type GameState = {
  gameId: string;
  gameName: string;
  isInGame: boolean;
  isInProgress: boolean;
  setIsInProgress: SetState<boolean>;
  setPlayers: SetState<Player[]>;
  setPlayerReady: (userId: string, isReady: boolean) => void;
  players: Player[];
};

const GameContext = createContext<GameState>({
  gameId: '',
  gameName: '',
  isInGame: false,
  isInProgress: false,
  setIsInProgress: () => undefined,
  setPlayerReady: () => undefined,
  setPlayers: () => undefined,
  players: [],
});

export const GameProvider: React.FC<{}> = ({ children }) => {
  const { socket } = useSocket();
  const { replace } = useHistory();
  const { pathname, state } = useLocation<GameLocationState>();
  const { gameId } = useParams<GameParams>();

  const [gameName] = useState(state?.gameInfo?.gameName ?? '');
  const [isInProgress, setIsInProgress] = useState(
    state?.gameInfo?.isInProgress ?? false
  );
  const [isInGame] = useState(state?.isInGame ?? false);
  const [players, setPlayers] = useState(state?.gameInfo?.players ?? []);

  const setPlayerReady = (userId: string, isReady: boolean): void => {
    setPlayers((prevPlayers) => {
      const index = prevPlayers.findIndex((player) => player.userId === userId);
      return [
        ...prevPlayers.slice(0, index),
        { ...prevPlayers[index], isReady },
        ...prevPlayers.slice(index + 1),
      ];
    });
  };

  useEffect(() => {
    if (isInGame) {
      replace(pathname);

      return (): void => {
        socket.emit('leaveGame', { gameId }, () => undefined);
      };
    }

    return;
  }, []);

  useEffect(() => {
    socket.on('playerJoined', (player: Player) => {
      setPlayers((prevPlayers) => [...prevPlayers, player]);
    });

    socket.on('playerLeft', ({ userId }: Player) =>
      setPlayers((prevPlayers) =>
        prevPlayers.filter((player) => player.userId !== userId)
      )
    );

    socket.on('playerReady', ({ userId }: Player) => {
      console.log(`${userId} is ready`);
      setPlayerReady(userId, true);
    });

    return (): void => {
      socket.off('playerJoined');
      socket.off('playerLeft');
      socket.off('playerReady');
    };
  }, []);

  return (
    <GameContext.Provider
      value={{
        gameId,
        gameName,
        isInGame,
        isInProgress,
        players,
        setIsInProgress,
        setPlayerReady,
        setPlayers,
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export const useGame = (): GameState => useContext(GameContext);
