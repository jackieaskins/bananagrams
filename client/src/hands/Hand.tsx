import React from 'react';
import Box from '@material-ui/core/Box';
import { useDrop } from 'react-dnd';

import { Hand as HandType } from './types';
import { useSocket } from '../SocketContext';
import Tile from '../tiles/Tile';
import { TileItem } from '../tiles/types';

type HandProps = {
  hand: HandType;
};

const Hand: React.FC<HandProps> = ({ hand }) => {
  const { socket } = useSocket();

  const [{ canDrop, isOver }, dropRef] = useDrop({
    accept: 'TILE',
    canDrop: ({ boardPosition }: TileItem, monitor) =>
      monitor.isOver() && !!boardPosition,
    drop: ({ boardPosition }: TileItem) => {
      socket.emit('moveTileFromBoardToHand', { boardPosition });
    },
    collect: (monitor) => ({
      isOver: monitor.isOver(),
      canDrop: monitor.canDrop(),
    }),
  });

  return (
    <Box
      display="flex"
      flexWrap="wrap"
      border={1}
      ref={dropRef}
      style={{
        backgroundColor: isOver && canDrop ? 'green' : 'transparent',
        marginTop: '10px',
        minHeight: '35px',
        opacity: isOver && canDrop ? 0.5 : 1,
        width: '525px',
      }}
    >
      {Object.values(hand).map((tile) => (
        <Tile key={tile.id} tile={tile} boardPosition={null} />
      ))}
    </Box>
  );
};

export default Hand;
