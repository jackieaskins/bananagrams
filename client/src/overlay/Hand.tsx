import { useMemo } from 'react';
import { useRecoilCallback } from 'recoil';

import { useCurrentHand } from '../game/stateHooks';
import Tile from '../tile/Tile';
import { selectedTileState } from '../tile/selectedTileState';

const Hand = (): JSX.Element => {
  const hand = useCurrentHand();

  const selectTile = useRecoilCallback(
    ({ set }) => ({ tile }) => {
      set(selectedTileState, { tile, boardPosition: null });
    },
    []
  );

  const tiles = useMemo(
    () =>
      (hand ?? []).map((tile) => (
        <Tile
          key={tile.id}
          id={tile.id}
          letter={tile.letter}
          onClick={selectTile}
        />
      )),
    [hand, selectTile]
  );

  return (
    <div
      className="overlay"
      css={{
        display: 'flex',
        gap: '5px',
        flexWrap: 'wrap',
        alignContent: 'flex-start',
      }}
    >
      {tiles}
    </div>
  );
};

export default Hand;
