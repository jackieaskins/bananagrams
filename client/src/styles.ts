import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

const DRAWER_WIDTH = 240;

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    drawer: {
      width: DRAWER_WIDTH,
    },
    drawerPaper: {
      width: DRAWER_WIDTH,
    },
    invalidDrop: {
      backgroundColor: theme.palette.error.main,
      color: theme.palette.error.contrastText,
      opacity: 0.5,
    },
    transparentPaper: {
      backgroundColor: 'transparent',
    },
    validDrop: {
      backgroundColor: theme.palette.success.main,
      color: theme.palette.success.contrastText,
      opacity: 0.5,
    },
  })
);
