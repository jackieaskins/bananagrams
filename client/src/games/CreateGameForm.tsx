import React from 'react';
import { Grid, Typography } from '@material-ui/core';

import Button from '../buttons/Button';
import ErrorAlert from '../alerts/ErrorAlert';
import TextField from '../forms/TextField';
import { useCreateGameForm } from './CreateGameFormState';

const CreateGameForm: React.FC = () => {
  const {
    error,
    gameName,
    isCreatingGame,
    onSubmit,
    setGameName,
    setUsername,
    username,
  } = useCreateGameForm();

  return (
    <>
      <Typography variant="h3" style={{ marginBottom: '5px' }}>
        Start a new game
      </Typography>
      <ErrorAlert visible={!!error} style={{ marginBottom: '5px' }}>
        {error}
      </ErrorAlert>
      <Grid
        container
        direction="column"
        spacing={3}
        component="form"
        autoComplete="off"
      >
        <Grid item>
          <TextField
            label="Game name"
            name="gameName"
            required
            margin="dense"
            value={gameName}
            setValue={setGameName}
          />

          <TextField
            label="Username"
            name="username"
            margin="dense"
            required
            value={username}
            setValue={setUsername}
          />
        </Grid>

        <Grid item>
          <Button
            loading={isCreatingGame}
            loadingText="Creating game"
            disabled={!gameName || !username}
            onClick={onSubmit}
            size="large"
            fullWidth
          >
            Create game
          </Button>
        </Grid>
      </Grid>
    </>
  );
};

export default CreateGameForm;
