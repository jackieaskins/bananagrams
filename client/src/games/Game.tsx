import { Grid, Typography } from '@material-ui/core';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';

import Board from '../boards/Board';
import OpponentBoardPreview from '../boards/OpponentBoardPreview';
import { isValidConnectedBoard } from '../boards/validate';
import Dump from '../hands/Dump';
import Hand from '../hands/Hand';
import { Player } from '../players/types';
import { useSocket } from '../socket/SocketContext';
import { useGame } from './GameContext';
import PeelButton from './PeelButton';

import './Game.css';

const Game = (): JSX.Element => {
  const { socket } = useSocket();
  const {
    gameInfo: { bunch, players },
    handlePeel,
  } = useGame();

  const { board, hand } = players.find(
    (player) => player.userId === socket.id
  ) as Player;

  const canPeel = hand.length === 0 && isValidConnectedBoard(board);
  const peelWinsGame = bunch.length < players.length;

  return (
    <DndProvider backend={HTML5Backend}>
      <Grid container spacing={2} justify="center" style={{ marginTop: '1px' }}>
        <Grid item>
          <Typography align="center" variant="body2" gutterBottom>
            Your board and hand:
          </Typography>

          <Grid container spacing={1} alignItems="center">
            <Grid item>
              <Board board={board} />
            </Grid>

            <Grid item>
              <Hand hand={hand} />
            </Grid>
          </Grid>
        </Grid>

        <Grid item>
          <Grid container direction="column" spacing={1} alignItems="center">
            <Grid item>
              <Typography variant="body2">
                Tiles remaining in bunch: {bunch.length}
              </Typography>
            </Grid>

            <Grid item style={{ width: '100%' }}>
              <PeelButton
                canPeel={canPeel}
                handlePeel={handlePeel}
                peelWinsGame={peelWinsGame}
              />
            </Grid>

            <Grid item style={{ width: '100%' }}>
              <Dump />
            </Grid>
            <Grid item>
              {players.length > 1 && (
                <Typography align="center" variant="body2">
                  Opponent board(s):
                </Typography>
              )}

              <OpponentBoardPreview players={players} />
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </DndProvider>
  );
};

export default Game;
