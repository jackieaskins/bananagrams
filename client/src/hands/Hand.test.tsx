import { Box, Button } from "@mui/material";
import { shallow } from "enzyme";
import { useDrop } from "react-dnd";
import { playerFixture } from "../fixtures/player";
import { useGame } from "../games/GameContext";
import { validDropSx } from "../styles";
import Hand from "./Hand";

const mockEmit = jest.fn();
jest.mock("../socket/SocketContext", () => ({
  useSocket: () => ({
    socket: { id: "123", emit: mockEmit },
  }),
}));

jest.mock("../styles", () => ({
  validDropSx: { color: "green" },
}));

jest.mock("../games/GameContext", () => ({
  useGame: jest.fn().mockReturnValue({
    gameInfo: { players: [] },
    handleMoveTileFromBoardToHand: jest.fn(),
  }),
}));

jest.mock("react-dnd", () => ({
  useDrop: jest
    .fn()
    .mockReturnValue([{ canDrop: true, isOver: false }, jest.fn()]),
}));

describe("<Hand />", () => {
  const hand = [
    { id: "A1", letter: "A" },
    { id: "B1", letter: "B" },
    { id: "C1", letter: "C" },
  ];

  const renderComponent = (propOverrides = {}) =>
    shallow(<Hand hand={hand} {...propOverrides} />);

  test("renders properly", () => {
    expect(renderComponent({ hand: [] })).toMatchSnapshot();
  });

  test("renders properly with player", () => {
    useGame.mockReturnValue({
      gameInfo: { players: [playerFixture({ userId: "123" })] },
      handleMoveTileFromBoardToHand: jest.fn(),
    });

    expect(renderComponent()).toMatchSnapshot();
  });

  test("shuffle button emits shuffle hand event", () => {
    renderComponent().find(Button).props().onClick();

    expect(mockEmit).toHaveBeenCalledWith("shuffleHand", {});
  });

  describe("useDrop", () => {
    const getUseDropCall = () => useDrop.mock.calls[0][0];

    test("is called", () => {
      renderComponent();
      expect(useDrop).toHaveBeenCalledWith({
        accept: "TILE",
        canDrop: expect.any(Function),
        drop: expect.any(Function),
        collect: expect.any(Function),
      });
    });

    describe("canDrop", () => {
      const canDrop = (tileItem, monitor) =>
        getUseDropCall().canDrop(tileItem, monitor);

      beforeEach(() => {
        renderComponent();
      });

      test("returns true if is over and tile is on board", () => {
        expect(
          canDrop(
            { boardLocation: { x: 0, y: 0 } },
            {
              isOver: () => true,
            },
          ),
        ).toBe(true);
      });

      test("returns false if is over but not on board", () => {
        expect(
          canDrop(
            { boardLocation: null },
            {
              isOver: () => true,
            },
          ),
        ).toBe(false);
      });

      test("returns false if is not over but is on board", () => {
        expect(
          canDrop(
            { boardLocation: { x: 0, y: 0 } },
            {
              isOver: () => false,
            },
          ),
        ).toBe(false);
      });

      test("returns false if is not over and is not on board", () => {
        expect(
          canDrop(
            { boardLocation: null },
            {
              isOver: () => false,
            },
          ),
        ).toBe(false);
      });
    });

    test("drop", () => {
      const mockHandleMoveTileFromBoardToHand = jest.fn();
      useGame.mockReturnValue({
        gameInfo: { players: [] },
        handleMoveTileFromBoardToHand: mockHandleMoveTileFromBoardToHand,
      });

      renderComponent();

      const boardLocation = { x: 0, y: 0 };
      getUseDropCall().drop({ boardLocation });

      expect(mockHandleMoveTileFromBoardToHand).toHaveBeenCalledWith(
        boardLocation,
      );
    });

    test("collect", () => {
      renderComponent();

      expect(
        getUseDropCall().collect({
          isOver: () => "isOver",
          canDrop: () => "canDrop",
        }),
      ).toEqual({ isOver: "isOver", canDrop: "canDrop" });
    });
  });

  describe("className", () => {
    test("has validDrop sx when is over and can drop", () => {
      useDrop.mockReturnValue([{ canDrop: true, isOver: true }]),
        expect(renderComponent().find(Box).props().sx).toEqual(validDropSx);
    });

    test("has no sx when is not over or cannot drop", () => {
      useDrop.mockReturnValue([{ canDrop: false, isOver: true }]),
        expect(renderComponent().find(Box).props().sx).toBeNull();
    });
  });
});
