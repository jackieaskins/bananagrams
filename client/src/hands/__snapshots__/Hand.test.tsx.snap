// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hand /> renders properly 1`] = `
<TransparentPaper>
  <ForwardRef(Tooltip)
    title="Shuffle hand"
  >
    <ForwardRef(Button)
      disabled={true}
      onClick={[Function]}
      size="small"
      sx={
        {
          "width": "100%",
        }
      }
    >
      <Memo(ForwardRef(ShuffleIcon))
        color="action"
        fontSize="small"
      />
    </ForwardRef(Button)>
  </ForwardRef(Tooltip)>
  <ForwardRef(Divider) />
  <ForwardRef(Box)
    display="flex"
    flexDirection="column"
    flexWrap="wrap"
    height="497px"
    p={1}
    sx={null}
  />
</TransparentPaper>
`;

exports[`<Hand /> renders properly with player 1`] = `
<TransparentPaper>
  <ForwardRef(Tooltip)
    title="Shuffle hand"
  >
    <ForwardRef(Button)
      disabled={false}
      onClick={[Function]}
      size="small"
      sx={
        {
          "width": "100%",
        }
      }
    >
      <Memo(ForwardRef(ShuffleIcon))
        color="action"
        fontSize="small"
      />
    </ForwardRef(Button)>
  </ForwardRef(Tooltip)>
  <ForwardRef(Divider) />
  <ForwardRef(Box)
    display="flex"
    flexDirection="column"
    flexWrap="wrap"
    height="-3px"
    p={1}
    sx={null}
  >
    <Tile
      boardLocation={null}
      key="A1"
      tile={
        {
          "id": "A1",
          "letter": "A",
        }
      }
    />
    <Tile
      boardLocation={null}
      key="B1"
      tile={
        {
          "id": "B1",
          "letter": "B",
        }
      }
    />
    <Tile
      boardLocation={null}
      key="C1"
      tile={
        {
          "id": "C1",
          "letter": "C",
        }
      }
    />
  </ForwardRef(Box)>
</TransparentPaper>
`;
